security/src/main/scala/pl/touk/nussknacker/ui/security/api/SecurityError.scala

```diff
@@ -1,18 +1,26 @@
 package pl.touk.nussknacker.ui.security.api
 
-sealed trait SecurityError
+sealed trait SecurityError {
+  val errorMessage: String
```

Comment:
That looks cool but If I am not mistaken it won't force a developer to provide string representation (errorMessage in this case)? There is a possibility that someone in the future will create `NewError extends SecurityError` without adding valid string representation to `Show[SecurityError]`.

Reply:
hmm, in this case, `Show[SecurityError]` should stop compiling (pattern matching exhaustivity check will detect that - at least when warns are treated as errors /in our case they are/)

Reply:
Ah, ok - so last thing I would be concerned about: Do we prefer to separate error message from case class/object definition to somewhere else and provide it through some import and use implicit method (with pattern matching inside) to provide `.show` method rather than just call variable from that class? 

Don't get me wrong - I love cats and I don't mind using implicits but I am wondering where is it beneficial and where is not :D

Reply:
TBH we don't have to use Show. We could have `def toString(error: SecurityError): String` method. This is just a single responsibility principle usage. In the core, you deal with some SecurityErrors. You declare them as ADT. Then at the end of the system (when you want to provide a RESTful interface) you need to change them to status codes and messages. In this the place where you say: this error maps to 403 and this one to 401. And this is the message related to the error.

So, this is not about Cats. It's rather about clean code ;)

Reply:
That is a good point and well explained :+1:

Reply:
I've used extension method to separate interpretations of errors from their definitions.